First version of the doc (#274)
ServiceNow/azimuth: docs/docs/user-guide/dataset-warnings.md

# Dataset Warnings

Datasets can suffer from a variety of issues, such as class imbalance, classes with low sample counts, and dataset shift. These warnings help detect some of these issues.

![](../_static/images/dataset-warnings/dataset-warnings-1.png)
![](../_static/images/dataset-warnings/dataset-warnings-2.png)
![](../_static/images/dataset-warnings/dataset-warnings-3.png)
![](../_static/images/dataset-warnings/dataset-warnings-4.png)

## Missing samples

In this first analysis, the application flags when a class has **fewer than `X`** (default is 20)
samples in either the training or the evaluation set. The plot helps to visualize the values for
each class.

## Class Imbalance

In this second analysis, Azimuth detects class imbalance issues. It raises a flag for all classes

> where the relative difference between the number of samples in that class and the mean sample count in a dataset split is above

Reviewer: ```suggestion
where the relative difference between the number of samples in that class and the mean sample count per class in a dataset split is above
```

a certain threshold `Y`. The default is 50%.

## Dataset Shift

A **discrepancy** between the training and evaluation splits can cause problems with a model. For
example, the model may **not have a representative sample** of examples to train on, making it **generalize poorly**
in production.

> Alternatively, you might be measuring the performance of the model on an evaluation set that may **
> not be a good indicator of the performance in production**. Distribution analysis aims to give

Reviewer: Oh...I just realized that this text was here and you just moved it. So here's a suggestion, and you can take it or leave it!

```suggestion
Alternatively, if your evaluation set does not come from the same data distribution as the data in production, measuring model performance on this evaluation set may **not be a good indicator of the performance in production**. Distribution analysis aims to give
```

# Dataset Warnings

> Dataset can suffer from different issues such as having classes with too few samples, class imbalance and dataset shift. These warnings help detect some of these issues.

Reviewer: ```suggestion
Datasets can suffer from a variety of issues, such as class imbalance, classes with low sample counts, and dataset shift. These warnings help detect some of these issues.
```

![](../_static/images/dataset-class-distribution-analysis/dataset-warnings-1.png)
![](../_static/images/dataset-class-distribution-analysis/dataset-warnings-2.png)
![](../_static/images/dataset-class-distribution-analysis/dataset-warnings-3.png)
TODO

## Missing samples

In this first analysis, the application flags when a class has **fewer than `X`** (default is 20)
samples in either the training or the evaluation set. The plot helps to visualize the values for
each class.

## Class Imbalance

> In this second analysis, Azimuth detects class imbalance issues. It raises a flag for all classes
> where the relative difference between the number of samples and the mean in a dataset split is above
> a certain threshold `Y`. The default is 50%.

Reviewer: ```suggestion
In this second analysis, Azimuth detects class imbalance issues. It raises a flag for all classes
where the relative difference between the number of samples in that class and the mean sample count in a dataset split is above
a certain threshold `Y`. The default is 50%.
```

Reviewer: This is the class imbalance flag we're adding, right? I'm feeling really confused because I don't understand how it was already in the docs. So if my suggestion doesn't make sense (the "sample count" part), ignore it!

Author: I did the documentation for class imbalance in that PR ;)

## Dataset Shift

> A **discrepancy** between the training and evaluation sets can cause problems with a model. For

Reviewer: ```suggestion
A **discrepancy** between the training and evaluation splits can cause problems with a model. For
```
I think?

> example, the model may **not have a representative sample** of examples to train on, making it **
> hard to generalize**

Reviewer: ```suggestion
example, the model may **not have a representative sample** of examples to train on, making it **generalize poorly**
```

in production.

Alternatively, you might be measuring the performance of the model on an evaluation set that may **

> not be a good indicator of the performance in production**. Distribution analysis aims to give

Reviewer: I like your phrasing in terms of "indicator," but I wonder if we should include the specific notion of "not representing production data" - or do you think that's implied?

Author: I don't get your meaning here.

Reviewer: So this is what I meant originally: 

I think there are two ways to look at the same problem, where one is focusing on the data itself, and the other is focusing on the interpretation / outcome:
1. The evaluation set does not come from the same data distribution as data in production, which is the same as it not representing production data
2. Measuring model performance on the evaluation set will not indicate performance in production

I like that you focused on the second, which is the real consequence, but since the problem is often described as the first, I wonder if we should include that, too. So something like: 

"Alternatively, if your evaluation set does not come from the same data distribution as the data in production, measuring model performance on this evaluation set may **not be a good indicator of the performance in production**.

But then I realized that that implies a difference between your eval set and actual production data that is...what? Not a dataset that you have access to?

So now I'm not actually sure what you mean by "may not be a good indicator of the performance in production" in the first place. Do you mean that the training data is a good representative, but the eval is not?

warnings when the training and evaluation sets look too different in some aspect of the data.

### Representation mismatch

> This analysis flags if a class is **over-represented** in the evaluation set (relative to

Reviewer: ```suggestion
This analysis raises a flag if a class is **over-represented** in the evaluation set (relative to
```
I always feel awkward about the use of "flag" in this way - I think it's because I expect a direct object? Like, "flags the class" or something? Anyway, it will still be understood the way it is, so your call.
